File: docs/JAMIA/comparisons/RPython/TM/30944577_str_contains_series_as_bool.ts
/**
 * RPython SO#30944577 — Check if string is in a pandas DataFrame
 * Effect: Crash
 * Bug class: Type coercion
 *
 * In pandas, str.contains() returns a Series of booleans. Using it in an
 * if-statement crashes because a Series is not a scalar boolean.
 *
 * In tidy-ts, filter() takes a predicate that operates on individual rows
 * and returns boolean. There is no Series/scalar confusion — the predicate
 * always returns a single boolean per row.
 */
import { createDataFrame, stats as s } from "@tidy-ts/dataframe";

const df = createDataFrame([
  { name: "Bob", births: 968 },
  { name: "Jessica", births: 155 },
  { name: "Mary", births: 77 },
  { name: "John", births: 578 },
  { name: "Mel", births: 973 },
]);

// filter predicate returns boolean per row — no ambiguity
const hasMel = df.filter((r) => r.name.includes("Mel"));
hasMel.print();

// Type system catches passing string column to numeric ops
// @ts-expect-error — string[] is not assignable to number[]
const wrong = s.mean(df.extract("name"));

File: docs/JAMIA/comparisons/RPython/TM/30944577_str_contains_series_as_bool.py
# ID: SO#30944577
# Language: Python
# Bug class: Value type
# Runtime consequence: Crash
# In study: Yes
# Inclusion rationale: str.contains returns Series used as scalar bool. Typed filter operates on values, returns boolean.
# Reproduction status: Live
# Type system catch: Typed accessor rejects wrong column type

import pandas as pd

df = pd.DataFrame({
    'Names': ['Bob', 'Jessica', 'Mary', 'John', 'Mel'],
    'Births': [968, 155, 77, 578, 973],
})

# BUG: str.contains returns Series, not scalar bool
try:
    if df['Names'].str.contains('Mel'):
        print("Mel is there")
except ValueError as e:
    print(f"Crash: {e}")
